core/src/main/java/com/lafaspot/sparzr/InternalSaxHandler.java

```diff
@@ -0,0 +1,355 @@
+                    ArrayNode jArray = (ArrayNode) currentItem.get(propertyName);
+                    jArray.add(new TextNode(content));
+                } else {
+                    final ArrayNode itemArray = jsonNodeFactory.arrayNode();
+                    itemArray.add(new TextNode(content));
+                    currentItem.set(propertyName, itemArray);
+                }
+                propertyName = null;
+                stringBuffer = new StringBuffer();
+            }
+            // If closing tag of an item, move up a level and return top-level item if necessary
+            final Format type = tagTypes.pop();
+            if (type.equals(Format.MICRODATA)) {
+                items.removeLast();
+                if (items.isEmpty() && currentItem != null) {
+                    for (final Listener listener : listeners) {
+                        listener.foundItem(currentItem);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Buffer characters from in-between tags.
+     */
+    @Override
+    public void characters(final char[] ch, final int start, final int length) throws SAXException {
+        if (containsJsonld || propertyName != null) {
+            stringBuffer.append(ch, start, length);
+        }
+    }
+
+    /**
+     * Recursively parse json to find all itemtypes.
+     *
+     * @param element Root json element
+     */
+    private void parseJson(final JsonNode element) {
+        if (element == null || element.isNull() || element.isValueNode() || element.isMissingNode()) {
+            return;
+        } else if (element.isObject()) {
+            final ObjectNode object = (ObjectNode) element;
+            final Iterator<Entry<String, JsonNode>> iterator = object.fields();
+            while (iterator.hasNext()) {
+                final Entry<String, JsonNode> entry = iterator.next();
+                final String key = entry.getKey();
+                final JsonNode value = entry.getValue();
+                if ("@type".equalsIgnoreCase(key)) {
+                    final String itemtype = value.textValue();
+                    foundItemtype(itemtype, Format.JSONLD);
+                } else {
+                    parseJson(value);
+                }
+            }
+        } else if (element.isArray()) {
+            final ArrayNode jsonArray = (ArrayNode) element;
+            final Iterator<JsonNode> jsonIterator = jsonArray.iterator();
+            while (jsonIterator.hasNext()) {
+                final JsonNode json = jsonIterator.next();
+                parseJson(json);
+            }
+        }
+    }
+
+    /**
+     * Alert listeners about a new item providing type and format information.
+     *
+     * @param type the specific itemtype
+     * @param format format of itemtype (Json-ld, Microdata etc)
+     */
+    private void foundItemtype(final String type, final Format format) {
```

Comment:
This should probably be a inner class of Listener.
Also rename Listener  to something less generic like SchemaListener.